### PR TITLE
[RFC] feat(slot): create generic slot component

### PR DESCRIPTION
**Pull request by bmdalex**
# Slot

This PR contains:
- implementation of generic `Slot` component;
- using it to convert `content` prop to shorthand slots

### TODO

- [x] Conformance test
- [x] Minimal doc site example
- [ ] Stardust base theme
- [x] Teams Light theme
- [ ] Teams Dark theme
- [ ] Teams Contrast theme
- [x] Confirm RTL usage
- [ ] [W3 accessibility](https://www.w3.org/standards/webdesign/accessibility) check
- [ ] [Stardust accessibility](https://github.com/stardust-ui/accessibility) check
- [ ] Update glossary props table
- [ ] Update the CHANGELOG.md

# API Proposal

## `content` prop as shorthand
We are now able to use `content` prop as real shorthand for our components if it's needed. Here are a couple of scenarios where this is useful:

### 2. Usage of `content` prop for `Button` component that was transformed to shorthand in this PR:

screen shot 2018-10-08 at 17 29 43

```jsx
<div>
  <Button content="Basic" />
  <Button content="Styled" styles={{ color: 'red', background: 'yellow' }} />
  <Button content={{ content: 'Shorthand styled', styles: { color: 'red', background: 'yellow' }}} />
</div>
```

```html
<div>
  <button class="ui-button" aria-disabled="false">
    <span class="ui-slot">Basic</span>
  </button>
  <button class="ui-button cf cj" aria-disabled="false">
    <span class="ui-slot">Styled</span>
  </button>
  <button class="ui-button" aria-disabled="false">
    <span class="ui-slot cf cj">Shorthand styled</span>
  </button>
</div>
```

### CODE REVIEW

File: src/components/Button/Button.tsx [hunk 1]
```diff
@@ -160,7 +161,9 @@ class Button extends UIComponent<Extendable<IButtonProps>, IButtonState> {
       >
         {hasChildren && children}
         {!hasChildren && iconPosition !== 'after' && this.renderIcon(variables, styles)}
-        {!hasChildren && content && <span className={classes.content}>{content}</span>}
+        {!hasChildren &&
+          content &&
+          Slot.create(content, { defaultProps: { as: 'span', className: classes.content } })}
```
Comment (levithomason) [inline src/components/Button/Button.tsx]: Short circuit checks not required for .create() methods.  The Slot will return nothing if content does not exist.

**Review by levithomason — APPROVED** [review]

File: src/components/Button/Button.tsx (hunk 1 above)
Comment (bmdalex) [reply src/components/Button/Button.tsx]: removed check

File: docs/src/examples/components/Slot/index.tsx [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import React from 'react'
+
+export default () => <div>No examples</div>
```
Comment (mnajdova) [inline docs/src/examples/components/Slot/index.tsx]: Maybe we can add here some description of the component and why it was introduced. This is not blocking this PR from being merged however.

**Review by mnajdova — APPROVED** [review]

File: docs/src/examples/components/Slot/index.tsx (hunk 2 above)
Comment (bmdalex) [reply docs/src/examples/components/Slot/index.tsx]: will do, thx

**Pull request merged**